Data/Scripts/001_Technical/002_RubyUtilities.rb

```diff
@@ -162,6 +159,147 @@ def self.move(source, destination)
+  # returns Hex color value as RGB
+  #-----------------------------------------------------------------------------
+  def self.hex_to_rgb(hex)
+  	hex = hex.to_s(16) if hex.is_a?(Numeric)
+  	r = hex[0...2].to_i(16)
+  	g = hex[2...4].to_i(16)
+  	b = hex[4...6].to_i(16)
+  	return r, g, b
+  end
+  #-----------------------------------------------------------------------------
+  # parse color input to return color object
+  #-----------------------------------------------------------------------------
+  def self.parse(color)
+    if color.is_a?(Color) # color object
+      return color
+    elsif color.is_a?(String) # string
+      if color.include?("#") # hex color
+        return Color.new(color)
+      elsif color.include(",") # RGB color
+        rgb = color.split(",").map(&:to_i)
+        return Color.new(*rgb)
+      end
+    elsif color.is_a?(Numeric) # decimal color
+      return Color.new(color)
+    end
+    # returns nothing if wrong input
+    return nil
+  end
+  #-----------------------------------------------------------------------------
+  # returns color object for some commonly used colors
+  #-----------------------------------------------------------------------------
+  def self.red; return Color.new(255, 0, 0); end
+  def self.green; return Color.new(0, 255, 0); end
+  def self.blue; return Color.new(0, 0, 255); end
+  def self.black; return Color.new(0, 0, 0); end
+  def self.white; return Color.new(255, 255, 255); end
+  def self.yellow; return Color.new(255, 255, 0); end
+  def self.orange; return Color.new(255, 155, 0); end
+  def self.purple; return Color.new(155, 0, 255); end
+  def self.brown; return Color.new(112, 72, 32); end
+  def self.teal; return Color.new(0, 255, 255); end
+  def self.magenta; return Color.new(255, 0, 255); end
+  #-----------------------------------------------------------------------------
+end
+
+#===============================================================================
+#  Wrap code blocks in class which passes data accessible as instance
+#  variables within the code block
+#
+#  wrapper = CallbackWrapper.new { puts @test }
+#  wrapper.set(test: 'Hi')
+#  wrapper.execute  #=>  'Hi'
+#===============================================================================
+class CallbackWrapper
+  @params = {}
+  #-----------------------------------------------------------------------------
+  #  constructor
+  #-----------------------------------------------------------------------------
+  def initialize(&block)
+    @code_block = block
+  end
+  #-----------------------------------------------------------------------------
+  #  execute callback
+  #-----------------------------------------------------------------------------
+  def execute(given_block = nil, *args)
+    execute_block = given_block || @code_block
+    @params.each do |key, value|
+      args.instance_variable_set("@#{key.to_s}", value)
+    end
```

Comment:
Shouldn't you make sure that none of the keys is called "params" or "code_block"? Perhaps raise it in `def set` for immediate feedback to the user that they're using a key name they shouldn't be.

I'm struggling to see the point of this class, to be honest. It just seems like a lambda but you can give it arguments separately from calling it (but you can only give it all the arguments at once via `set`, rather than giving it one and then another), and I don't know where that would be useful. Maybe I'm misreading this.

Reply:
One use case I can think of for this is replacing the procs in `BattleHandlers` where common arguments for the procs there, like `battle`, `battler`, `ability` and `item` can be set beforehand for the CallbackWrapper and can be accessed within the proc without explicitly passing it as an argument.

Reply:
☝️ This. I'm using the same approach for things like animation code definition in EBDX. It allows the system to expose a slew of good to have instance variables to the code block, without the need for explicitly defining the same set of arguments for each of them (especially when it comes to arguments that are shared across code blocks).

Reply:
Additionally, the names of the variables passed doesn't matter. They're expressed as symbols - there won't be any funky interaction:
```

    wrapper = CallbackWrapper.new do
      echoln "Test block output"
      echoln @params
      echoln @code_block
    end
    
    wrapper.set(params: 1, code_block: 2)
    wrapper.execute
    
```
Prints out
```
Test block output
1
2
```